pcdshub/lcls-twincat-common-components
# ENH: temperature ffos

## Description

Apply Maggie's pmps database temperatures to her temperature ffos in the context of the common components.

Related links:
- https://github.com/pcdshub/lcls-twincat-pmps/blob/master/lcls-twincat-pmps/PMPS/AncillaryComponents/FB_TempSensor_FFO.TcPOU
- https://github.com/pcdshub/lcls-twincat-general/blob/master/LCLSGeneral/LCLSGeneral/POUs/Hardware/FB_TempSensor.TcPOU

More detailed info:

- Create a helper subclass of the temp sensor pmps block called `FB_CC_TempSensor`
- Replace all temperature handling with the new subclass
- Set fault thresholds to the current setpoint's temp setting when possible
- Veto it in the out state
- Include provisions for the SATT (2-4 blade stack style), which previously did not have pmps db states
- Include provisions for power slits, which previously did not have pmps db states
- For the slits, use the FBs from lcls-plc-kfe-motion, which are strictly more useful than the versions from lcls-plc-lfe-motion.
- Add tests for everything I added
- Add reset logic on veto + no autoreset

## Motivation and Context


We can prevent the beam from melting certain devices with temperature measurements.
The temperatures can be stored in the database.

Jira https://jira.slac.stanford.edu/browse/ECS-4472

## How Has This Been Tested?



- I wrote tests for every device and the base FB
- The tests are largely structured as:

  - Set the threshold via spoofing the PMPS DB in memory (start from https://github.com/pcdshub/lcls-twincat-motion/blob/master/lcls-twincat-motion/Library/Tests/Helpers/FB_PMPSJsonTestHelper.TcPOU and expand)
  - Set the IO readback temperature to way above or way below the threshold via pointer cheating
  - Confirm they react accordingly at various states while the states mover test moves things around

## Where Has This Been Documented?


This needs to be announced on release and on deploy. Deployment must be done carefully to avoid rolling over the FFO slots cap on each PLC.

## Pre-merge checklist
- [x] Code contains descriptive comments
- [x] Libraries are set to ``Always Newest`` version (``Library, *``)
- [x] Committed with ``pre-commit`` or ran ``pre-commit run --all-files``
